docs/decisions/0017-reimplement-asset-processing.rst

```diff
@@ -0,0 +1,268 @@
+     - Better parameterize the input and output paths edx-platform asset build, such that it may `search for node_modules outside of edx-platform <https://github.com/openedx/wg-developer-experience/issues/150> and `generate assets outside of edx-platform <https://github.com/openedx/wg-developer-experience/issues/151>`.
+
+All of these potential solutions would involve refactoring or entirely replacing parts of the current asset processing system.
+
+
+Decision
+********
+
+We will largely reimplement edx-platform's asset processing system. We will aim to:
+
+* Use well-known, npm-installed frontend tooling wherever possible.
+* When bespoke processing is required, use standard POSIX tools like Bash.
+* When Django/Python is absolutely required, contain its impact so that the rest of the system remains Python-free.
+* Avoid unnecessary indirection or abstraction. For this task, extensibility is a non-goal, and simplicity is a virtue.
+* Provide a clear migration path from the old system to the new one.
+* Enable the future removal of as much legacy frontend tooling code as possible.
+
+Consequences
+************
+
+Reimplementation Specification
+==============================
+
+The three top-level edx-platform asset processing actions are *build*, *collect*, and *watch*. The build action can be further broken down into five stages. Here is how those actions and stages will be reimplemented:
+
+
+.. list-table::
+   :header-rows: 1
+
+   * - Description
+     - Old implementation
+     - New implementation
+
+   * - **Build: All stages.** Compile, generate, copy, and otherwise process static assets so that they can be used by the Django webserver or collected elsewhere. For many Web applications, all static asset building would be coordinated via Webpack or another NPM-managed tool. Due to the age of edx-platform and its legacy XModule and Comprehensive Theming systems, though, there are five stages which need to be performed in a particular order.
+
+     - ``paver update_assets --skip-collect``
+
+       A Python-defined task that calls out to each build stage.
+
+     - ``assets/build.sh``
+
+       A Bash script that contains all build stages, with subcommands available for running each stage separately. Its command-line interface inspired by Tutor's ``openedx-assets`` script. The script will be runnable on any POSIX system, including macOS and Ubuntu and it will linted for common shell scripting mistakes using `shellcheck <https://www.shellcheck.net>`_.
+     
+   * - + **Build stage 1: Copy npm-installed assets** from node_modules to other folders in edx-platform. They are used by certain especially-old legacy LMS & CMS frontends that are not set up to work with npm directly.
+
+     - ``paver update_assets --skip-collect``
+
+       Implemented in Python within update_assets. There is no standalone command for it.
+
+     - ``assets/build.sh npm``
```

Comment:
My (admittedly limited) understanding of npm and webpack was that these components should be in charge of collecting assets from `node_module/` to the user's directory. For instance, assets could (should?) be automatically copied whenever we run `npm intall` thanks to a "postinstall" script in package.json: https://docs.npmjs.com/cli/v6/using-npm/scripts#pre--post-scripts This approach would mean that we no longer need to run asset collection for npm dependencies.

Reply:
Noted.